File: typescript/ch2/Student.ts
import { Person_fp, Person_ip } from './Person';

class Student_ip extends Person_ip {
  _school: string;

  constructor(firstname: string, lastname: string, ssn: string, school: string) {
    super(firstname, lastname, ssn);
    this._school = school;
  }

  get school() {
    return this._school;
  }

  studentsInSameCountryAndSchool(friends: Student_ip[]) {
    const closeFriends = super.peopleInSameCountry(friends);
    const result = [];
    for (const i in closeFriends) {
      const friend = closeFriends[i] as Student_ip;
      if (this.school == friend.school) {
        result.push(friend);
      }
    }

    return result;
  }
}

class Student_fp extends Person_fp {
  _school: string;

  constructor(firstname: string, lastname: string, ssn: string, school: string) {
    super(firstname, lastname, ssn);
    this._school = school;
  }

  get school() {
    return this._school;
  }
}

export {
  Student_ip,
  Student_fp
};

File: python/python/ch2/Student.py
from .Person import Person_ip, Person_fp

class Student_ip(Person_ip):
    def __init__(self, firstname, lastname, ssn, school):
        super().__init__(firstname, lastname, ssn)
        self.school = school
        
    def __str__(self):
        return f'{super().__str__()}, {self.school}'
        
    def students_in_same_country_and_school(self, friends):
        close_friends = super().people_in_same_country(friends)
        result = []
        for friend in close_friends:
            if self.school == friend.school:
                result.append(str(friend))
                
        return result
      
class Student_fp(Person_fp):
    def __init__(self, firstname, lastname, ssn, school):
        super().__init__(firstname, lastname, ssn)
        self.school = school
        
    def __str__(self):
        return f'{super().__str__()}, {self.school}'